import into failed with tidb log error“read xxx bytes from external storage, exceed max limit 1073741824” when inject network latency form s3 (global sort dir)

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1、open global sort and tidb_enable_dist_task
2、import into
3、inject network latency form s3 (global sort dir)

### 2. What did you expect to see? (Required)
import into success

### 3. What did you see instead (Required)

the status of import job is not finished or running (now: 2023-12-16 04:32:37, jobId: 150003, step: importing, status: failed)
operatorLogs:
[2023-12-16 04:06:07] ###### start import into
[2023-12-16 04:06:07] ###### wait for import job to finish
[2023-12-16 04:32:37] ###### wait for import job to finish failed
select id, step, status from mysql.tidb_import_jobs where start_time >= '2023-12-16 04:06:07'
jobId: 150003, step: importing, status: failed

tidb logs:
import into failed with log error“read xxx bytes from external storage, exceed max limit 1073741824”

[2023/12/16 04:32:06.313 +08:00] [ERROR] [reader.go:48] ["read all data failed"] [takeTime=1m2.635809773s] [error="read 3772936486386853641 bytes from external storage, exceed max limit 1073741824"]
[2023/12/16 04:32:06.313 +08:00] [ERROR] [local.go:1710] ["do import meets error"] [error="read 3772936486386853641 bytes from external storage, exceed max limit 1073741824"]

### 4. What is your TiDB version? (Required)
git hash：
074af1082d5a15019cb2720f54e942374ea83d1a

--- a/br/pkg/lightning/backend/external/byte_reader.go
+++ b/br/pkg/lightning/backend/external/byte_reader.go
@@ -147,7 +147,8 @@ func (r *byteReader) switchConcurrentMode(useConcurrent bool) error {
 	// is it's the end of file. When it's the end of the file, caller will see EOF
 	// and no further switchConcurrentMode should be called.
 	largeBufSize := readerFields.bufSizePerConc * readerFields.concurrency
-	delta := int64(offsetInOldBuf + (reloadCnt-1)*largeBufSize)
+	delta := int64(offsetInOldBuf + (reloadCnt-1)*largeBufSize + (r.curBufIdx)*readerFields.bufSizePerConc)
+
 	if _, err := r.storageReader.Seek(delta, io.SeekCurrent); err != nil {
 		return err
 	}
@@ -210,6 +211,9 @@ func (r *byteReader) readNBytes(n int) ([]byte, error) {
 	// need to flatten bs
 	if n > int(size.GB) {
 		return nil, errors.Errorf("read %d bytes from external storage, exceed max limit %d", n, size.GB)
+	}
+	if n < 0 {
+		return nil, errors.Errorf("read %d bytes from external storage", n)
 	}
 	auxBuf := make([]byte, n)
 	for _, b := range bs {
@@ -314,6 +318,7 @@ func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) 
 	r.logger.Info("drop data in closeConcurrentReader",
 		zap.Int("reloadCnt", r.concurrentReader.reloadCnt),
 		zap.Int("dropBytes", len(r.curBuf[r.curBufIdx])-r.curBufOffset),
+		zap.Int("curBufIdx", r.curBufIdx),
 	)
 	r.concurrentReader.largeBufferPool.Destroy()
 	r.concurrentReader.largeBuf = nil

--- a/br/pkg/lightning/backend/external/byte_reader.go
+++ b/br/pkg/lightning/backend/external/byte_reader.go
@@ -313,7 +313,7 @@ func (r *byteReader) reload() error {
 func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) {
 	r.logger.Info("drop data in closeConcurrentReader",
 		zap.Int("reloadCnt", r.concurrentReader.reloadCnt),
-		zap.Int("dropBytes", len(r.curBuf)-r.curBufOffset),
+		zap.Int("dropBytes", len(r.curBuf[r.curBufIdx])-r.curBufOffset),
 	)
 	r.concurrentReader.largeBufferPool.Destroy()
 	r.concurrentReader.largeBuf = nil

--- a/br/pkg/lightning/backend/external/byte_reader.go
+++ b/br/pkg/lightning/backend/external/byte_reader.go
@@ -317,7 +317,7 @@ func (r *byteReader) reload() error {
 func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) {
 	r.logger.Info("drop data in closeConcurrentReader",
 		zap.Int("reloadCnt", r.concurrentReader.reloadCnt),
-		zap.Int("dropBytes", len(r.curBuf[r.curBufIdx])-r.curBufOffset),
+		zap.Int("dropBytes", r.concurrentReader.bufSizePerConc*(len(r.curBuf)-r.curBufIdx)-r.curBufOffset),
 		zap.Int("curBufIdx", r.curBufIdx),
 	)
 	r.concurrentReader.largeBufferPool.Destroy()

--- a/br/pkg/lightning/backend/external/byte_reader.go
+++ b/br/pkg/lightning/backend/external/byte_reader.go
@@ -212,8 +212,8 @@ func (r *byteReader) readNBytes(n int) ([]byte, error) {
 	if n > int(size.GB) {
 		return nil, errors.Errorf("read %d bytes from external storage, exceed max limit %d", n, size.GB)
 	}
-	if n < 0 {
-		return nil, errors.Errorf("read %d bytes from external storage", n)
+	if n <= 0 {
+		return nil, errors.Errorf("illegal n (%d) when reading from external storage", n)
 	}
 	auxBuf := make([]byte, n)
 	for _, b := range bs {

--- a/br/pkg/lightning/backend/external/byte_reader.go
+++ b/br/pkg/lightning/backend/external/byte_reader.go
@@ -147,7 +147,7 @@ func (r *byteReader) switchConcurrentMode(useConcurrent bool) error {
 	// is it's the end of file. When it's the end of the file, caller will see EOF
 	// and no further switchConcurrentMode should be called.
 	largeBufSize := readerFields.bufSizePerConc * readerFields.concurrency
-	delta := int64(offsetInOldBuf + (reloadCnt-1)*largeBufSize + (r.curBufIdx)*readerFields.bufSizePerConc)
+	delta := int64(offsetInOldBuf + (reloadCnt-1)*largeBufSize)
 
 	if _, err := r.storageReader.Seek(delta, io.SeekCurrent); err != nil {
 		return err
@@ -326,7 +326,7 @@ func (r *byteReader) closeConcurrentReader() (reloadCnt, offsetInOldBuffer int) 
 	reloadCnt = r.concurrentReader.reloadCnt
 	r.concurrentReader.reloadCnt = 0
 	r.curBuf = [][]byte{r.smallBuf}
-	offsetInOldBuffer = r.curBufOffset
+	offsetInOldBuffer = r.curBufOffset + r.curBufIdx*r.concurrentReader.bufSizePerConc
 	r.curBufOffset = 0
 	return
 }

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -209,7 +209,7 @@ func TestEmptyContent(t *testing.T) {
 func TestSwitchMode(t *testing.T) {
 	st := storage.NewMemStorage()
 	// Prepare
-	fileSize := 1024 * 1024
+	fileSize := 1024 * 1024 * 40
 	err := st.WriteFile(context.Background(), "/testfile", make([]byte, fileSize))
 	require.NoError(t, err)
 
@@ -219,14 +219,14 @@ func TestSwitchMode(t *testing.T) {
 		return rsc
 	}
 	pool := membuf.NewPool()
-	ConcurrentReaderBufferSizePerConc = 100
-	br, err := newByteReader(context.Background(), newRsc(), 100)
-	require.NoError(t, err)
-	br.enableConcurrentRead(st, "/testfile", 100, 100, pool.NewBuffer())
-
+	// ConcurrentReaderBufferSizePerConc = rand.Intn(300) + 1
 	seed := time.Now().Unix()
 	rand.Seed(uint64(seed))
 	t.Logf("seed: %d", seed)
+	br, err := newByteReader(context.Background(), newRsc(), 64*1024)
+	require.NoError(t, err)
+	br.enableConcurrentRead(st, "/testfile", 100, ConcurrentReaderBufferSizePerConc, pool.NewBuffer())
+
 	totalCnt := 0
 	modeUseCon := false
 	for totalCnt < fileSize {
@@ -239,7 +239,7 @@ func TestSwitchMode(t *testing.T) {
 				modeUseCon = true
 			}
 		}
-		n := rand.Intn(100)
+		n := rand.Intn(300)
 		if n == 0 {
 			n = 1
 		}

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -29,6 +29,7 @@ import (
 	"github.com/johannesboyne/gofakes3/backend/s3mem"
 	"github.com/pingcap/errors"
 	backuppb "github.com/pingcap/kvproto/pkg/brpb"
+	"github.com/pingcap/tidb/br/pkg/lightning/common"
 	"github.com/pingcap/tidb/br/pkg/membuf"
 	"github.com/pingcap/tidb/br/pkg/storage"
 	"github.com/stretchr/testify/require"
@@ -207,56 +208,62 @@ func TestEmptyContent(t *testing.T) {
 }
 
 func TestSwitchMode(t *testing.T) {
-	st := storage.NewMemStorage()
-	// Prepare
-	fileSize := 1024 * 1024 * 40
-	err := st.WriteFile(context.Background(), "/testfile", make([]byte, fileSize))
-	require.NoError(t, err)
-
-	newRsc := func() storage.ExternalFileReader {
-		rsc, err := st.Open(context.Background(), "/testfile", nil)
-		require.NoError(t, err)
-		return rsc
-	}
-	pool := membuf.NewPool()
-	// ConcurrentReaderBufferSizePerConc = rand.Intn(300) + 1
 	seed := time.Now().Unix()
 	rand.Seed(uint64(seed))
 	t.Logf("seed: %d", seed)
-	br, err := newByteReader(context.Background(), newRsc(), 64*1024)
-	require.NoError(t, err)
-	br.enableConcurrentRead(st, "/testfile", 100, ConcurrentReaderBufferSizePerConc, pool.NewBuffer())
+	st := storage.NewMemStorage()
+	// Prepare
+	ctx := context.Background()
+	writer := NewWriterBuilder().
+		SetPropSizeDistance(100).
+		SetPropKeysDistance(2).
+		BuildOneFile(st, "/test", "0")
 
-	totalCnt := 0
+	err := writer.Init(ctx, 5*1024*1024)
+	require.NoError(t, err)
+
+	kvCnt := 1000000
+	kvs := make([]common.KvPair, kvCnt)
+	for i := 0; i < kvCnt; i++ {
+		randLen := rand.Intn(10) + 1
+		kvs[i].Key = make([]byte, randLen)
+		_, err := rand.Read(kvs[i].Key)
+		require.NoError(t, err)
+		randLen = rand.Intn(10) + 1
+		kvs[i].Val = make([]byte, randLen)
+		_, err = rand.Read(kvs[i].Val)
+		require.NoError(t, err)
+	}
+
+	for _, item := range kvs {
+		err := writer.WriteRow(ctx, item.Key, item.Val)
+		require.NoError(t, err)
+	}
+
+	err = writer.Close(ctx)
+	require.NoError(t, err)
+	pool := membuf.NewPool()
+	ConcurrentReaderBufferSizePerConc = rand.Intn(300) + 1
+	kvReader, err := newKVReader(context.Background(), "/test/0/one-file", st, 0, 64*1024)
+	require.NoError(t, err)
+	kvReader.byteReader.enableConcurrentRead(st, "/test/0/one-file", 100, ConcurrentReaderBufferSizePerConc, pool.NewBuffer())
 	modeUseCon := false
-	for totalCnt < fileSize {
-		if rand.Intn(5) == 0 {
+	for {
+		if rand.Intn(2) == 0 {
 			if modeUseCon {
-				br.switchConcurrentMode(false)
+				kvReader.byteReader.switchConcurrentMode(false)
 				modeUseCon = false
 			} else {
-				br.switchConcurrentMode(true)
+				kvReader.byteReader.switchConcurrentMode(true)
 				modeUseCon = true
 			}
 		}
-		n := rand.Intn(300)
-		if n == 0 {
-			n = 1
-		}
-		if totalCnt+n > fileSize {
-			n = fileSize - totalCnt
-		}
-		if n == 0 {
-			break
-		}
-		y, err := br.readNBytes(n)
+		_, _, err := kvReader.nextKV()
 		if err == io.EOF {
 			break
 		}
 		require.NoError(t, err)
-		totalCnt += len(y)
 	}
-	require.Equal(t, fileSize, totalCnt)
 }
 
 // NewS3WithBucketAndPrefix creates a new S3Storage for testing.

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -209,6 +209,7 @@ func TestEmptyContent(t *testing.T) {
 
 func TestSwitchMode(t *testing.T) {
 	seed := time.Now().Unix()
+	// seed := 1702880814
 	rand.Seed(uint64(seed))
 	t.Logf("seed: %d", seed)
 	st := storage.NewMemStorage()
@@ -243,7 +244,7 @@ func TestSwitchMode(t *testing.T) {
 	err = writer.Close(ctx)
 	require.NoError(t, err)
 	pool := membuf.NewPool()
-	ConcurrentReaderBufferSizePerConc = rand.Intn(300) + 1
+	ConcurrentReaderBufferSizePerConc = rand.Intn(100) + 1
 	kvReader, err := newKVReader(context.Background(), "/test/0/one-file", st, 0, 64*1024)
 	require.NoError(t, err)
 	kvReader.byteReader.enableConcurrentRead(st, "/test/0/one-file", 100, ConcurrentReaderBufferSizePerConc, pool.NewBuffer())

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -209,7 +209,6 @@ func TestEmptyContent(t *testing.T) {
 
 func TestSwitchMode(t *testing.T) {
 	seed := time.Now().Unix()
-	// seed := 1702880814
 	rand.Seed(uint64(seed))
 	t.Logf("seed: %d", seed)
 	st := storage.NewMemStorage()

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -29,6 +29,7 @@ import (
 	"github.com/johannesboyne/gofakes3/backend/s3mem"
 	"github.com/pingcap/errors"
 	backuppb "github.com/pingcap/kvproto/pkg/brpb"
+	"github.com/pingcap/tidb/br/pkg/membuf"
 	"github.com/pingcap/tidb/br/pkg/storage"
 	"github.com/stretchr/testify/require"
 	"golang.org/x/exp/rand"
@@ -206,22 +207,22 @@ func TestEmptyContent(t *testing.T) {
 }
 
 func TestSwitchMode(t *testing.T) {
-	st, clean := NewS3WithBucketAndPrefix(t, "test", "testprefix")
-	defer clean()
-
+	st := storage.NewMemStorage()
 	// Prepare
 	fileSize := 1024 * 1024
-	err := st.WriteFile(context.Background(), "testfile", make([]byte, fileSize))
+	err := st.WriteFile(context.Background(), "/testfile", make([]byte, fileSize))
 	require.NoError(t, err)
 
 	newRsc := func() storage.ExternalFileReader {
-		rsc, err := st.Open(context.Background(), "testfile", nil)
+		rsc, err := st.Open(context.Background(), "/testfile", nil)
 		require.NoError(t, err)
 		return rsc
 	}
-
+	pool := membuf.NewPool()
 	ConcurrentReaderBufferSizePerConc = 100
 	br, err := newByteReader(context.Background(), newRsc(), 100)
+	require.NoError(t, err)
+	br.enableConcurrentRead(st, "/testfile", 100, 100, pool.NewBuffer())
 
 	seed := time.Now().Unix()
 	rand.Seed(uint64(seed))
@@ -256,7 +257,6 @@ func TestSwitchMode(t *testing.T) {
 		totalCnt += len(y)
 	}
 	require.Equal(t, fileSize, totalCnt)
-
 }
 
 // NewS3WithBucketAndPrefix creates a new S3Storage for testing.

--- a/br/pkg/lightning/backend/external/byte_reader_test.go
+++ b/br/pkg/lightning/backend/external/byte_reader_test.go
@@ -248,8 +248,9 @@ func TestSwitchMode(t *testing.T) {
 	require.NoError(t, err)
 	kvReader.byteReader.enableConcurrentRead(st, "/test/0/one-file", 100, ConcurrentReaderBufferSizePerConc, pool.NewBuffer())
 	modeUseCon := false
+	i := 0
 	for {
-		if rand.Intn(2) == 0 {
+		if rand.Intn(5) == 0 {
 			if modeUseCon {
 				kvReader.byteReader.switchConcurrentMode(false)
 				modeUseCon = false
@@ -258,11 +259,14 @@ func TestSwitchMode(t *testing.T) {
 				modeUseCon = true
 			}
 		}
-		_, _, err := kvReader.nextKV()
+		key, val, err := kvReader.nextKV()
 		if err == io.EOF {
 			break
 		}
 		require.NoError(t, err)
+		require.Equal(t, kvs[i].Key, key)
+		require.Equal(t, kvs[i].Val, val)
+		i++
 	}
 }
 

